### ISSUE TITLE
[Feature Request] Mass User Management & Existing User Detection

### ISSUE THREAD

**Issue by JcfeJohn** [hrfee/jfa-go#5]
**Mass User Management**
-

* Able to mass update the settings of All Users (both Admin and regular users), Regular users, and Admin users.

Settings that can be updated are:
* Personal Preferences
* Library Access
* Possibly everything in **Profile** section

This is useful if you have handful of users.

Existing User Detection
-

* When someone tries to register an existing user, it will prompt "Username is already in used"

**Comment by hrfee** [comment]
I like the sound of this, the basic functionality shouldn't be too hard but I'd like to do a decent UI. I Might merge the email and user defaults menu in with this as well.

Also, the existing user detection should already be implemented. I don't ever remember testing it though so it could just be bugged.

**Comment by JcfeJohn** [comment]
I just tested the existing user detection, it still goes through and create the account.

Also, upon testing the forgot password, I have encountered this error:

```
[INFO] 17:02:02 New password reset for user "test02"
panic: interface conversion: interface {} is nil, not string

goroutine 35 [running]:
main.pwrMonitor(0xc00013b680, 0xc0000b28c0)
        /home/hrfee/projects/go/jfa-go/pwreset.go:70 +0xcc5
created by main.(*appContext).StartPWR
        /home/hrfee/projects/go/jfa-go/pwreset.go:29 +0x1d3
```

Meanwhile, upon account creation, I get this thing:

```
[ERROR] 17:00:25 api.go:230: qrMiurBYiFPgTN7Mggtw3H: Failed to send user creation notification
```

Inviting user gives this error:

```
[ERROR] 17:05:29 api.go:317: bqMXdyPxJg6HzrXmMzewQo: Failed to send to bl4k15@tempmail.top
```

**Comment by hrfee** [comment]
Ok, i thought i fixed that but it looks like jf.userByName isn't working for you then. You downloaded the new version right?

**Comment by JcfeJohn** [comment]
Yes, I downloaded the latest version 0.1.8

**Comment by hrfee** [comment]
And for the email problem, could you restart with the `-debug` flag and try again? there should be another line printed below the "Failed to send to x" message with a more descriptive error.

**Comment by JcfeJohn** [comment]
Here it is:
```
[DEBUG] 17:14:16 api.go:288: Generating new invite
[DEBUG] 17:14:16 api.go:308: 2kRQP4PwpMDJcJcwFYWt9P: Sending invite email
[ERROR] 17:14:17 api.go:317: 2kRQP4PwpMDJcJcwFYWt9P: Failed to send to bl4k15@tempmail.top
[DEBUG] 17:14:17 api.go:318: 2kRQP4PwpMDJcJcwFYWt9P: Error: 535 Incorrect authentication data
[GIN-debug] [WARNING] Headers were already written. Wanted to override status code 400 with 200
[GIN/DEBUG] 17:14:17: POST(/generateInvite) => 200 in 695.542953ms;
```

I'm using the same SMTP credentials of the old jf-accounts. I just tested and it is working fine there but not on jfa-go

**Comment by hrfee** [comment]
For the password reset bug, try [this](https://nextcloud.hrfee.pw/s/eSkNokMnfBcJY3D) build. Run `jfa-go test`, and when it asks to view output, enter 'n'. It will then ask for a username, enter 'test02' or whichever account name. It should print the status code and error, as well as the user details.

For the email problem, could you send the [email] and [smtp] sections of your configuration? go ahead and censor your email/password if you need to.

**Comment by JcfeJohn** [comment]
The email magically worked. I just switched to mailgun (filled up the mailgun credentials) then switched back to smtp. It is now sending emails.

Going to test the build now

**Comment by hrfee** [comment]
thats weird. hopefully it won't happen again.

**Comment by JcfeJohn** [comment]
```
jfa-go version: 0.1.8 (802f957)
Socket: /tmp/jfa-go.sock
[INFO] 17:35:22 Authenticated with http://localhost:8096


----

access token : XXXXXXXXXXX
username : jf-accounts
server : http://localhost:8096
server version : 10.6.4
server name : Movies Central
authenticated? : true
response encoding: gzip
getUsers: code 200 err %!s(<nil>) maplength 38
View output? [y/n]: n
Enter a user to grab: test03
userByName (test03): code 200 err %!s(<nil>){
  "Configuration": {
    "DisplayCollectionsView": false,
    "DisplayMissingEpisodes": false,
    "EnableLocalPassword": false,
    "EnableNextEpisodeAutoPlay": true,
    "GroupedFolders": [],
    "HidePlayedInLatest": true,
    "LatestItemsExcludes": [],
    "MyMediaExcludes": [],
    "OrderedViews": [],
    "PlayDefaultAudioTrack": true,
    "RememberAudioSelections": true,
    "RememberSubtitleSelections": true,
    "SubtitleLanguagePreference": "",
    "SubtitleMode": "Default"
  },
  "EnableAutoLogin": false,
  "HasConfiguredEasyPassword": true,
  "HasConfiguredPassword": true,
  "HasPassword": true,
  "Id": "39a9742ed1bf40f6a93876d60331d739",
  "Name": "test03",
  "Policy": {
    "AccessSchedules": [],
    "AuthenticationProviderId": "Jellyfin.Server.Implementations.Users.DefaultAuthenticationProvider",
    "BlockUnratedItems": [],
    "BlockedChannels": [],
    "BlockedMediaFolders": [],
    "BlockedTags": [],
    "EnableAllChannels": false,
    "EnableAllDevices": true,
    "EnableAllFolders": false,
    "EnableAudioPlaybackTranscoding": true,
    "EnableContentDeletion": false,
    "EnableContentDeletionFromFolders": [],
    "EnableContentDownloading": true,
    "EnableLiveTvAccess": false,
    "EnableLiveTvManagement": false,
    "EnableMediaConversion": true,
    "EnableMediaPlayback": true,
    "EnablePlaybackRemuxing": true,
    "EnablePublicSharing": true,
    "EnableRemoteAccess": true,
    "EnableRemoteControlOfOtherUsers": false,
    "EnableSharedDeviceControl": false,
    "EnableSyncTranscoding": true,
    "EnableUserPreferenceAccess": true,
    "EnableVideoPlaybackTranscoding": true,
    "EnabledChannels": [],
    "EnabledDevices": [],
    "EnabledFolders": [
      "70f8a014f1004adeb77e1b644acc307b",
      "da8c836392781d2a9b70516c52f1f8bb",
      "a27ea59c6e20d4f874743014dce23bd3",
      "d97d363bc5b66994dc7b26181d42681e",
      "9d7ad6afe9afa2dab1a2f6e00ad28fa6"
    ],
    "ForceRemoteSourceTranscoding": false,
    "InvalidLoginAttemptCount": 0,
    "IsAdministrator": false,
    "IsDisabled": false,
    "IsHidden": true,
    "LoginAttemptsBeforeLockout": -1,
    "PasswordResetProviderId": "Jellyfin.Server.Implementations.Users.DefaultPasswordResetProvider",
    "RemoteClientBitrateLimit": 0,
    "SyncPlayAccess": "CreateAndJoinGroups"
  },
  "ServerId": "3115a1a4ed7e485eb3b9939bb165ae95"
}
```

**Comment by hrfee** [comment]
well thats working. I made another change for the build i sent but I didn't expect it to fix anything. Try removing the `test` flag, add the `-debug` one again and do a password reset again.

**Comment by JcfeJohn** [comment]
It's working now

```
[INFO] 17:39:12 New password reset for user "test03"
[INFO] 17:39:14 Sent password reset email to "bl4k15@tempmail.top"
```

Thanks!

**Comment by hrfee** [comment]
great. use that build for now until I release the next version i guess. I'll start working on the new features now.

**Comment by JcfeJohn** [comment]
Great! Thanks again, really appreciate it

**Comment by hrfee** [comment]
also, could you try the existing user detection again? It used the same function that wasn't working for password resets, so it should be fixed now too.

**Comment by JcfeJohn** [comment]
It is working now

```
[INFO] 18:33:05 qrMiurBYiFPgTN7Mggtw3H New user failed: User already exists named Carl
```

**Comment by hrfee** [comment]
On the feature request, it seems like all i can provide options for is access rights and basic playback options (everything you can see in your [comment](https://github.com/hrfee/jfa-go/issues/5#issuecomment-693554713)), and homescreen layout. Also, the feature will work similarly to the current user defaults, as in you create a template user and the program will copy the settings to other users.

**Comment by JcfeJohn** [comment]
There seems to be a problem with the build. When creating an account, it is stuck with **Creating** then upon checking the CPU usage of the server, Jellyfin increases the CPU usage.

![image](https://user-images.githubusercontent.com/31156280/93382260-77624a80-f894-11ea-9fca-96d30ec675ba.png)

The downhill was when I restarted the Jellyfin server.

I did ran with `-debug` flag but it doesn't show anything else beside New user attempt.

I tried to create an account again and it increased the CPU usage of Jellyfin again, stuck in Creating, then upon stopping jfa-go, CPU usage dropped.

It's quite late here now (3:20 am) and I'm off to bed. My apologies again

**Comment by hrfee** [comment]
Yeah sorry, I noticed the problem after I sent it to you, I accidentally made `userByName` loop it couldn't find the user. The latest commit fixes it.

**Comment by hrfee** [comment]
Here's a little demo of what i've got so far (excuse my terrible censoring):

![demo](https://dump.hrfee.pw/demo.gif)

As you can see, you can select a bunch of users and then apply settings to them. You can either use your stored user template, or grab the configuration from a chosen user.

Later on I'll put a little edit icon next to the users email addresses and get rid of the menu for that in settings. Anything else I should add?

Also, right now, you click on 'invites' or 'accounts' at the top to access each tab. I think its kinda ugly, any ideas for it?

**Comment by JcfeJohn** [comment]
Sorry for the delayed response.

Looks awesome!

Since there is an "Add User", maybe add "Delete User"?

I think it looks cool (minimalistic). Unfortunately, I have no idea for that

**Comment by hrfee** [comment]
Ok, ill add a delete button.

**Comment by JcfeJohn** [comment]
Oh, maybe add an email feature to "Delete User" too?

Once pressed, it will prompt "Do you want to send emails?"

Pressed "No" -> proceeds to account deletion

Pressed "Yes" -> ask for "Reason" -> Sends an email to the user if they have an email registered 

```Accounts - Jellyfin
Hello XXXX,

Your account has been deleted from our server.


(if there is a reason provided)
Reason:
XXXXXXX

(no reason provided)
No reason provided.
```

**Comment by hrfee** [comment]
I like that a lot. Will add.

**Comment by hrfee** [comment]
Just pushed everything, I think its pretty much done. [here's](https://nextcloud.hrfee.pw/s/csbtwTF3B2PfCEA) a build to try out.

**Comment by JcfeJohn** [comment]
Looks great! Using it right now and haven't encountered any issue

**Comment by JcfeJohn** [comment]
Getting this error:

```
interface conversion: interface {} is nil, not string
/usr/lib/go/src/runtime/iface.go:261 (0x40af2e)
/home/hrfee/projects/go/jfa-go/api.go:457 (0xa6d7d0)
/home/hrfee/projects/gopath/pkg/mod/github.com/gin-gonic/gin@v1.6.3/context.go:161 (0x9899fa)
/home/hrfee/projects/go/jfa-go/auth.go:103 (0xa74e17)
/home/hrfee/projects/gopath/pkg/mod/github.com/gin-gonic/gin@v1.6.3/context.go:161 (0x9899fa)
/home/hrfee/projects/gopath/pkg/mod/github.com/gin-gonic/gin@v1.6.3/recovery.go:83 (0x99d6c4)
/home/hrfee/projects/gopath/pkg/mod/github.com/gin-gonic/gin@v1.6.3/context.go:161 (0x9899fa)
/home/hrfee/projects/gopath/pkg/mod/github.com/gin-gonic/gin@v1.6.3/logger.go:241 (0x99c7c4)
/home/hrfee/projects/gopath/pkg/mod/github.com/gin-gonic/gin@v1.6.3/context.go:161 (0x9899fa)
/home/hrfee/projects/gopath/pkg/mod/github.com/gin-gonic/gin@v1.6.3/gin.go:409 (0x993a19)
/home/hrfee/projects/gopath/pkg/mod/github.com/gin-gonic/gin@v1.6.3/gin.go:367 (0x99310c)
/usr/lib/go/src/net/http/server.go:2843 (0x706b42)
/usr/lib/go/src/net/http/server.go:1925 (0x70224c)
/usr/lib/go/src/runtime/asm_amd64.s:1374 (0x46faa0)
```

![image](https://user-images.githubusercontent.com/31156280/93661853-d63ee400-fa8d-11ea-95c8-862f14ec404d.png)

**Comment by hrfee** [comment]
Do you have an email setup for the account you log into jfa-go with?

**Comment by JcfeJohn** [comment]
Oh, I didn't noticed that. It's working now. Thanks and sorry for the inconvenience!

**Comment by hrfee** [comment]
still a bug, i'll fix it.

**Comment by hrfee** [comment]
Released in v0.1.9
